src/features/fileOpenCloseProvider.ts

```diff
@@ -0,0 +1,74 @@
+import { IDisposable } from "../Disposable";
+import { OmniSharpServer } from "../omnisharp/server";
+import * as vscode from 'vscode';
+import CompositeDisposable from "../CompositeDisposable";
+import * as serverUtils from '../omnisharp/utils';
+import { isVirtualCSharpDocument } from "./virtualDocumentTracker";
+
+export default function fileOpenClose(server: OmniSharpServer): IDisposable {
+    return new FileOpenCloseProvider(server);
+}
+
+class FileOpenCloseProvider implements IDisposable {
+    private _server: OmniSharpServer;
+    private _diagnostics: vscode.DiagnosticCollection;
+    private _disposable: CompositeDisposable;
+
+    constructor(server: OmniSharpServer) {
+        this._server = server;
+        this._diagnostics = vscode.languages.createDiagnosticCollection('csharp');
+
+        setTimeout(async () => {
+            for (let editor of vscode.window.visibleTextEditors) {
+                let document = editor.document;
+
+                await this._onDocumentOpen(document);
+            }
+        }, 0);
+
+        this._disposable = new CompositeDisposable(this._diagnostics,
+            vscode.workspace.onDidOpenTextDocument(this._onDocumentOpen, this),
+            vscode.workspace.onDidCloseTextDocument(this._onDocumentClose, this),
+            vscode.window.onDidChangeActiveTextEditor(this._onActiveTextEdtiorChnage, this)
+        );
+    }
+
+    private async _onDocumentOpen(e: vscode.TextDocument) {
+        if (shouldIgnoreDocument(e)) {
+            return;
+        }
+
+        await serverUtils.fileOpen(this._server, { FileName: e.fileName });
+    }
+
+    private async _onDocumentClose(e: vscode.TextDocument) {
+        if (shouldIgnoreDocument(e)) {
+            return;
+        }
+
+        await serverUtils.fileClose(this._server, { FileName: e.fileName });
+    }
+
+    private async _onActiveTextEdtiorChnage(e: vscode.TextEditor) {
+        if (shouldIgnoreDocument(e.document)) {
+            return;
+        }
+
+        await serverUtils.filesChanged(this._server, [{ FileName: e.document.fileName }]);
+    }
+
+    dispose = () => this._disposable.dispose();
+}
+
+function shouldIgnoreDocument(document: vscode.TextDocument) {
+    if (document.languageId !== 'csharp') {
+        return true;
+    }
+
+    if (document.uri.scheme !== 'file' &&
+        !isVirtualCSharpDocument(document)) {
```

Comment:
This probably should check source-generated files as well?

Reply:
So it does check those files, I think that this just reads a little weird (it's the same logic that in `diagnosticsProvider.ts`). So if the scheme is file or it is a virtual document then it won't be ignored (though it's phrased as the inverse.)

Reply:
I don't believe so. Virtual documents have to do with razor remapping, and I don't think we want to ignore roslyn analyzers for razor files. cc @NTaylorMullen for input.

Reply:
Yep, what I'm saying is that it isn't ignoring the razor virtual files:
![image](https://user-images.githubusercontent.com/804724/157767013-f3ca9841-2b95-4773-bd6b-0613649fdb6a.png)

So that check is the same as the one in `diagnosticsProvider.ts`, and the open / close is currently to control how analyzers run, so open/close should be ignoring the same files that `diagnosticsProvider.ts` is, right?

Reply:
I think I misread this condition the first time. This is fine.